distrobuilder/vm.go

```diff
@@ -0,0 +1,146 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	lxd "github.com/lxc/lxd/shared"
+
+	"github.com/lxc/distrobuilder/shared"
+)
+
+var loopDevice string
+
+func getRootfsDevFile() string {
+	if loopDevice == "" {
+		return ""
+	}
+
+	return filepath.Join("/dev", "mapper", fmt.Sprintf("%sp1", loopDevice))
+}
+
+func getUEFIDevFile() string {
+	if loopDevice == "" {
+		return ""
+	}
+
+	return filepath.Join("/dev", "mapper", fmt.Sprintf("%sp15", loopDevice))
+}
+
+func createEmptyDiskImage(imageFile string, size uint) error {
+	return shared.RunCommand("dd", "if=/dev/zero", fmt.Sprintf("of=%s", imageFile), "count=0", fmt.Sprintf("seek=%d", size))
+}
+
+func createPartitions(imageFile string) error {
+	args := [][]string{
+		{"--zap-all"},
+		{"--new=15::+100M", "--new=1::"},
+		{"-t 15:ef00"},
+	}
+
+	for _, cmd := range args {
+		err := shared.RunCommand("sgdisk", append([]string{imageFile}, cmd...)...)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func mountImage(imageFile string) error {
+	// If loopDevice is set, it probably is already mounted.
+	if loopDevice != "" {
+		return nil
+	}
+
+	var stdout bytes.Buffer
+
+	err := lxd.RunCommandWithFds(nil, &stdout, "kpartx", "-a", "-s", "-v", imageFile)
```

Comment:
Can you see if you can't avoid `kpartx` altogether?
It's pretty annoying to run inside containers, so if you can instead use `losetup` with a calculated offset, that would be much more compatible with running distrobuilder in jenkins.